Content.Client/ContextMenu/UI/EntityMenuPresenter.cs

```diff
@@ -0,0 +1,421 @@
+                OpenRootMenu(entities);
+                return true;
+            }
+
+            //visibility checks
+            player.TryGetContainer(out var playerContainer);
+            foreach (var entity in entities.ToList())
+            {
+                if (!entity.TryGetComponent(out ISpriteComponent? spriteComponent) ||
+                    !spriteComponent.Visible ||
+                    !CanSeeContainerCheck(entity, playerContainer))
+                {
+                    entities.Remove(entity);
+                }
+            }
+
+            if (entities.Count == 0)
+                return false;
+
+            OpenRootMenu(entities);
+            return true;
+        }
+
+        /// <summary>
+        ///     Can the player see the entity through any entity containers?
+        /// </summary>
+        /// <remarks>
+        ///     This is similar to <see cref="ContainerHelpers.IsInSameOrParentContainer()"/>, except that we do not
+        ///     allow the player to be the "parent" container and we allow for see-through containers (display cases). 
+        /// </remarks>
+        private bool CanSeeContainerCheck(IEntity entity, IContainer? playerContainer)
+        {
+            // is the player inside this entity?
+            if (playerContainer?.Owner == entity)
+                return true;
+
+            entity.TryGetContainer(out var entityContainer);
+
+            // are they in the same container (or none?)
+            if (playerContainer == entityContainer)
+                return true;
+
+            // Is the entity in a display case?
+            if (playerContainer == null && entityContainer!.ShowContents)
+                return true;
+
+            return false;
+        }
+
+        /// <summary>
+        ///     Check that entities in the context menu are still visible. If not, remove them from the context menu.
+        /// </summary>
+        public void Update()
+        {
+            if (!RootMenu.Visible)
+                return;
+
+            var player = _playerManager.LocalPlayer?.ControlledEntity;
+
+            if (player == null)
+                return;
+
+            foreach (var entity in Elements.Keys.ToList())
+            {
+                if (entity.Deleted)
+                {
+                    RemoveEntity(entity);
+                    continue;
+                }
+
+                // Do we need to do in-range unOccluded checks?
+                if (_verbSystem.CanSeeAllContext || !_eyeManager.CurrentEye.DrawFov)
+                    continue;
+
+                if (!player.InRangeUnOccluded(entity))
+                    RemoveEntity(entity);
+            }
+        }
+
+        /// <summary>
+        ///     Add menu elements for a list of grouped entities; 
+        /// </summary>
+        /// <param name="entityGroups"> A list of entity groups. Entities are grouped together based on prototype.</param>
+        private void AddToUI(List<List<IEntity>> entityGroups)
```

Comment:
double list is sus but i guess it makes sense

Reply:
other than renaming context menu -> entity menu, I made no real changes to `EntityMenuPresenterGrouping.cs` which has all the entity-grouping logic and stores them as these nested lists. It's probably fine as it is though